### ISSUE TITLE
[실험] Qwen3-8B LoRA 파인튜닝 - 데이터 스케일링 실험 (162→477→958 samples)

### ISSUE THREAD

**Issue by ryangkyung-thaki** [ThakiCloud/LLaMA-Factory#14]
## 📋 실험 개요
Qwen3-8B 모델의 LoRA fine-tuning에서 학습 데이터 크기에 따른 성능 변화 분석

## 🎯 실험 목적
- 인터뷰 데이터셋 크기 증가에 따른 학습 효과 검증
- 최적의 데이터셋 크기 결정
- LoRA 파라미터와 데이터 크기의 관계 분석

### 공통 LoRA 구성
- **lora_rank**: 16
- **lora_alpha**: 32
- **lora_dropout**: 0.05
- **lora_target**: q_proj, k_proj, v_proj, o_proj, gate_proj, up_proj, down_proj (7 modules)
- **trainable_params**: ~61.3M
- **bias**: none

### 공통 학습 설정
- **Model**: Qwen3-8B (36 layers, d_model=4096)
- **per_device_train_batch_size**: 4
- **gradient_accumulation_steps**: 8
- **effective_batch_size**: 32
- **learning_rate**: 2e-5
- **optimizer**: AdamW
- **lr_scheduler**: cosine
- **warmup_ratio**: 0.03
- **weight_decay**: 0.01
- **num_train_epochs**: 3.0
- **cutoff_len**: 10,000 tokens

### Experiment 1: Baseline (162 samples)
- **Train samples**: 162
- **Eval samples**: 43
- **Steps per epoch**: 6
- **Total steps**: 18 (6 × 3 epochs)
- **Training time**: ~23분 28초
- **Avg time per step**: 62.56초/step
- **Final train_loss**: 1.0207

### Experiment 2: 3배 확장 (477 samples)
- **Train samples**: 477 (+315, +194%)
- **Eval samples**: 123 (+80, +186%)
- **Steps per epoch**: 15
- **Total steps**: 45 (15 × 3 epochs)
- **Training time**: ~1시간 9분 (4,156초)
- **Avg time per step**: 92.32초/step
- **Final train_loss**: 0.8238

### Experiment 3: 6배 확장 (958 samples)
- **Train samples**: 958 (+796 from baseline, +493%)
- **Eval samples**: 242 (+199 from baseline, +463%)
- **Steps per epoch**: 30
- **Total steps**: 90 (30 × 3 epochs)
- **Training time**: 진행 중
- **Expected time**: ~2-3시간

### 1. Loss 개선
- Exp1 → Exp2: **1.0207 → 0.8238** (-19.3% 개선)
- 데이터 3배 증가로 loss 약 20% 감소

### 2. 학습 시간
- Exp1: 23분 28초 (18 steps)
- Exp2: 1시간 9분 (45 steps, 2.5배 증가)
- Step당 시간 증가: 62.56초 → 92.32초 (+47.6%)

### 3. LoRA 파라미터-데이터 비율

| Experiment | Train Tokens* | Param-Token Ratio | LoRA 이론 적정 범위 |
|-----------|---------------|-------------------|---------------------|
| Exp1 | ~1M | 61.3:1 | ❌ 데이터 부족 (61M 필요) |
| Exp2 | ~3.1M | 19.8:1 | ⚠️ 여전히 부족 |
| Exp3 | ~6.2M | 9.9:1 | ⚠️ 개선 중 (92M 권장) |

*추정치 (평균 6,500 tokens/sample 기준)

### Baseline 실험 (162 samples, 18 steps)
```yaml
eval_strategy: epoch
save_steps: 6  # epoch마다
```

### 600 samples 실험 (477 samples, 45 steps)
```yaml
eval_strategy: epoch
save_steps: 15  # epoch마다
```

### 1200 samples 실험 (958 samples, 90 steps)
```yaml
eval_strategy: steps
eval_steps: 15  # 반 epoch마다 (6번 평가)
save_strategy: steps
save_steps: 30  # epoch마다
```

### LoRA 파라미터 계산 방식
```
각 LoRA adapter = 2 × d_model × rank (A, B 행렬)
전체 trainable params = num_target_modules × 2 × d_model × rank

Qwen3-8B:
- Attention (Q,K,V,O): 4 × (2 × 4096 × 16) = 524,288 params/layer
- MLP (gate,up,down): 3 × (2 × 12288 × 16) = 1,179,648 params/layer
- Per layer: 1,703,936 params
- Total (36 layers): 61.3M trainable parameters
```

### Step 계산 방식
```
Effective Batch Size = per_device_batch × gradient_accumulation × num_gpus
Steps per epoch = num_samples ÷ effective_batch_size
Total steps = steps_per_epoch × num_epochs

Example (Exp3):
- 958 samples ÷ 32 batch = 30 steps/epoch
- 30 × 3 epochs = 90 total steps
```

### Token 요구량 (LoRA 연구 기반)
- **Minimum (1:1 ratio)**: trainable_params × 1 = 61M tokens
- **Optimal (1.5:1 ratio)**: trainable_params × 1.5 = 92M tokens
- **Current (Exp3)**: ~6.2M tokens → 더 많은 데이터 필요

### 1. Evaluation 결과 분석
- [ ] checkpoint-15, 30, 45 (Exp3) 평가 완료 대기
- [ ] KMMLU, KoBEST, HAE-RAE 벤치마크 비교
- [ ] Loss curve 분석

### 2. 추가 실험 고려사항
- [ ] 더 큰 데이터셋 (1500-2000 samples) 테스트
- [ ] lora_rank 조정 실험 (8, 16, 32 비교)
- [ ] cutoff_len 최적화 (현재 10K tokens)

### 3. 최적화 포인트
- [ ] 학습 속도 개선 (step당 시간 단축)
- [ ] 메모리 사용량 최적화
- [ ] 최적 checkpoint 선정

## 📁 관련 파일
- Config: `examples/train_lora/qwen3_8b.yaml`
- Training logs:
  - `qwen3_8b_interview.log` (162 samples)
  - `qwen3_8b_interview_600_data.log` (477 samples)
  - `qwen3_8b_interview_1200_data.log` (958 samples)
- Checkpoints: `saves/qwen3-8b/lora/sft/`

## 🔗 참고 자료
- LoRA vs Full Fine-tuning: 데이터-파라미터 비율의 중요성
- Qwen3-8B Model Card
- LLaMA-Factory Documentation

## 📝 메모
- 데이터 증가에 따른 loss 개선 확인됨
- 하지만 LoRA 이론적 최적 데이터량에는 여전히 부족
- 실제 downstream task 성능 측정 필요
